cpp/src/arrow/util/bit-util.cc

```diff
@@ -108,47 +108,85 @@ int64_t CountSetBits(const uint8_t* data, int64_t bit_offset, int64_t length) {
 }
 
 template <bool invert_bits>
-Status TransferBitmap(MemoryPool* pool, const uint8_t* data, int64_t offset,
-                      int64_t length, std::shared_ptr<Buffer>* out) {
-  std::shared_ptr<Buffer> buffer;
-  RETURN_NOT_OK(AllocateEmptyBitmap(pool, length, &buffer));
-  uint8_t* dest = buffer->mutable_data();
-
+void TransferBitmap(const uint8_t* data, int64_t offset, int64_t length,
+                    int64_t dest_offset, uint8_t* dest) {
   int64_t byte_offset = offset / 8;
   int64_t bit_offset = offset % 8;
+  int64_t dest_byte_offset = dest_offset / 8;
+  int64_t dest_bit_offset = dest_offset % 8;
   int64_t num_bytes = BitUtil::BytesForBits(length);
-  int64_t bits_to_zero = num_bytes * 8 - length;
+  // Shift dest by its byte offset
+  dest += dest_byte_offset;
 
-  if (bit_offset > 0) {
-    uint32_t carry_mask = BitUtil::kBitmask[bit_offset] - 1U;
-    uint32_t carry_shift = 8U - static_cast<uint32_t>(bit_offset);
-
-    uint32_t carry = 0U;
-    if (BitUtil::BytesForBits(length + bit_offset) > num_bytes) {
-      carry = (data[byte_offset + num_bytes] & carry_mask) << carry_shift;
+  if (dest_bit_offset > 0) {
+    for (int64_t i = 0; i < length; i++) {
+      if (BitUtil::GetBit(data, i + offset)) {
+        BitUtil::SetBit(dest, i + dest_bit_offset);
+      } else {
+        BitUtil::ClearBit(dest, i + dest_bit_offset);
+      }
+    }
+  } else {
+    // Take care of the trailing bits in the last byte
+    int64_t trailing_bits = num_bytes * 8 - length;
+    uint8_t trail = 0;
+    if (trailing_bits) {
+      trail = dest[num_bytes - 1];
     }
 
-    int64_t i = num_bytes - 1;
-    while (i + 1 > 0) {
-      uint8_t cur_byte = data[byte_offset + i];
+    if (bit_offset > 0) {
+      uint32_t carry_mask = BitUtil::kBitmask[bit_offset] - 1U;
+      uint32_t carry_shift = 8U - static_cast<uint32_t>(bit_offset);
+
+      uint32_t carry = 0U;
+      if (BitUtil::BytesForBits(length + bit_offset) > num_bytes) {
+        carry = (data[byte_offset + num_bytes] & carry_mask) << carry_shift;
+      }
+
+      int64_t i = num_bytes - 1;
+      while (i + 1 > 0) {
+        uint8_t cur_byte = data[byte_offset + i];
+        if (invert_bits) {
+          dest[i] = static_cast<uint8_t>(~((cur_byte >> bit_offset) | carry));
+        } else {
+          dest[i] = static_cast<uint8_t>((cur_byte >> bit_offset) | carry);
+        }
+        carry = (cur_byte & carry_mask) << carry_shift;
+        --i;
```

Comment:
I would expect a copy loop to proceed in ascending order, except to avoid overlap (which isn't a concern here). I'm not sure doing it in descending order has any adversarial consequence, but it looks a bit weird.